### PR TITLE
Added functionality to select a flow in an alternatives list

### PR DESCRIPTION
**Pull request by AlistairDoswald**
- also corrected several bugs

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java [hunk 1]
```diff
@@ -153,10 +154,42 @@ public Response processAction(String actionExecution) {
...
+            for (AuthenticationExecutionModel flowExecution: flowExecutions) {
+                recursiveClearExecutionStatus(flowExecution);
+            }
+        }
+    }
+
+    /**
+     * This method makes sure that the parent flow's corresponding execution is considered successful if its contained
+     * executions are successful.
+     * The purpose is for when an execution is validated through an action, to make sure it's parent flow can be successful
```
Comment (fperot74) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java]: Typo: its

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java [hunk 2]
```diff
@@ -153,10 +154,42 @@ public Response processAction(String actionExecution) {
...
+     * when re-evaluation the flow tree.
+     *
+     * @param model An execution model.
+     */
+    private void checkAndValidateParentFlow(AuthenticationExecutionModel model) {
+        List<AuthenticationExecutionModel> localExecutions = processor.getRealm().getAuthenticationExecutions(model.getParentFlow());
+        AuthenticationExecutionModel parentFlowModel = processor.getRealm().getAuthenticationExecutionByFlowId(model.getParentFlow());
+        if ((model.isRequired() && localExecutions.stream().allMatch(m -> processor.isSuccessful(m))) ||
```
Comment (fperot74) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java]: m -> processor.isSuccessful(m) should be replaced by processor::isSuccessful

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java [hunk 3]
```diff
@@ -153,10 +154,42 @@ public Response processAction(String actionExecution) {
...
     }
 
+    /**
+     * Removes the execution status for an execution. If it is a flow, do the same for all sub-executions.
+     * @param execution the execution for which the status must be cleared
+     */
+    private void recursiveClearExecutionStatus(AuthenticationExecutionModel execution) {
+        processor.getAuthenticationSession().getExecutionStatus().remove(execution.getId());
+        if (execution.isAuthenticatorFlow()) {
+            List<AuthenticationExecutionModel> flowExecutions = processor.getRealm().getAuthenticationExecutions(execution.getFlowId());
```
Comment (fperot74) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java]: Can be shorten:
processor.getRealm().getAuthenticationExecutions(execution.getFlowId()).forEach(this::recursiveClearExecutionStatus);

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java [hunk 4]
```diff
@@ -314,6 +347,15 @@ private Response processSingleFlowExecutionModel(AuthenticationExecutionModel mo
         return processResult(context, false);
     }
 
+    /**
+     * This method creates the the list of authenticators that is presented to the user. For a required execution, this is
```
Comment (fperot74) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java]: **the the** list

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java [hunk 5]
```diff
@@ -332,6 +374,8 @@ private Response processSingleFlowExecutionModel(AuthenticationExecutionModel mo
                         }
                         CredentialValidator<?> cv = (CredentialValidator<?>) localAuthenticator;
                         typeAuthExecMap.put(cv.getType(processor.getSession()), execution);
+                    } else {
+                        nonCredentialExecutions.add(execution);
```
Comment (fperot74) [inline services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java:375]: Don't we need to check if it is an instance of CredentialValidator (or something like this) ?

**Review by fperot74 — APPROVED** [review]

File: services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java (hunk 5 above)
Comment (AlistairDoswald) [reply services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java:375]: no, this addition specifically adds flows for the non-credential authentication options.

**Pull request merged**